### PR TITLE
Implement triggers

### PR DESCRIPTION
**Pull request by jnwatson**
Triggers cause storm queries to be automatically made when certain conditions occur.

### CODE REVIEW

File: synapse/cmds/trigger.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    event happens.
+
+    Syntax:
+        trigger.list
+
+    Example:
+        trigger.list
+        (FIXME example output)
+    '''
+    _cmd_name = 'trigger.list'
```
[the remark this answers is not in the document]
Comment (jnwatson) [reply synapse/cmds/trigger.py]: I took the conversation to heart to :) Let's talk about it.

File: synapse/common.py [hunk 2]
```diff
@@ -680,6 +680,7 @@ def result(retn):
 
     ctor = getattr(s_exc, name, None)
     if ctor is not None:
+        # breakpoint()
```
[the remark this answers is not in the document]
Comment (jnwatson) [reply synapse/common.py]: yeah I caught that in my next pass

File: synapse/cmds/trigger.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+            self.printf(f'{user:10} {buidf:12} {cond:9} {obj:14} {obj2:10} {query}')
+
+async def _match_buids(self, core, prefix):
+    '''
+    Returns the buid that starts with prefix.  Prints out error and returns None if it doesn't match
+    exactly one.
+    '''
+    prefix = prefix.encode()
+    buids = [buid for buid, trig in await core.listTriggers()]
+    matches = [buid for buid in buids if binascii.hexlify(buid).startswith(prefix)]
```
[the remark this answers is not in the document]
Comment (jnwatson) [reply synapse/cmds/trigger.py]: ok

File: synapse/cortex.py [hunk 4]
```diff
@@ -102,6 +102,41 @@ def getModelDict(self):
...
+        Adds a trigger to the cortex
+        '''
+        username = None if self.user is None else self.user.name
+        self.cell.triggers.add(username, condition, query, form=form, tag=tag, prop=prop)
+
+    async def delTrigger(self, buid):
+        '''
+        Deletes aa trigger from the cortex
+        '''
+        trig = self.cell.triggers.find(buid)
```
[the remark this answers is not in the document]
Comment (jnwatson) [reply synapse/cortex.py]: ok

File: synapse/datamodel.py [hunk 5]
```diff
@@ -422,22 +429,23 @@ def getPropsByType(self, name):
         # TODO order props based on score...
         return props
 
-    def _addTypeDecl(self, decl):
+    # FIXME:  isn't used and won't work.  DELETE ME
```
[the remark this answers is not in the document]
Comment (jnwatson) [reply synapse/datamodel.py]: ok

File: synapse/cortex.py [hunk 6]
```diff
@@ -102,6 +102,41 @@ def getModelDict(self):
         '''
         return self.cell.model.getModelDict()
 
+    async def addTrigger(self, condition, query, *, form=None, tag=None, prop=None):
+        '''
+        Adds a trigger to the cortex
+        '''
+        username = None if self.user is None else self.user.name
+        self.cell.triggers.add(username, condition, query, form=form, tag=tag, prop=prop)
```
[the remark this answers is not in the document]
Comment (jnwatson) [reply synapse/cortex.py]: ok

File: synapse/cortex.py [hunk 7]
```diff
@@ -102,6 +102,41 @@ def getModelDict(self):
...
+            raise s_exc.AuthDeny(user=self.user.name, mesg='May only delete triggers created by you')
+        self.cell.triggers.delete(buid)
+
+    async def updateTrigger(self, buid, query):
+        '''
+        Change an existing trigger's query
+        '''
+        trig = self.cell.triggers.find(buid)
+        username = None if self.user is None else self.user.name
+        if username is not None and (not self.user.admin) and username != trig.get('user'):
```
[the remark this answers is not in the document]
Comment (jnwatson) [reply synapse/cortex.py]: ok

File: synapse/cortex.py [hunk 8]
```diff
@@ -992,15 +1039,21 @@ def getStormQuery(self, text):
         '''
         Parse storm query text and return a Query object.
         '''
-        return s_syntax.Parser(self, text).query()
+        parseinfo = {
+            'stormcmds': list(self.stormcmds.keys()),
```
[the remark this answers is not in the document]
Comment (jnwatson) [reply synapse/cortex.py]: ok

File: synapse/cortex.py [hunk 9]
```diff
@@ -1046,6 +1099,13 @@ def _logStormQuery(self, text, user):
             async for node in snap.getNodesBy(full, valu, cmpr=cmpr):
                 yield node
 
+    def getCoreInfo(self):
```
Comment (vEpiphyte) [inline synapse/cortex.py]: This is getting very close to duplicating the `stat()` functionality.
Suggestion: Move this data into the stat() results, remove the adminapi helper around the coreApi.stat() information.

File: synapse/cortex.py [hunk 10]
```diff
@@ -996,15 +1043,21 @@ def getStormQuery(self, text):
...
+        return query
 
     def _logStormQuery(self, text, user):
         '''
         Log a storm query.
         '''
         if self.conf.get('storm:log'):
             lvl = self.conf.get('storm:log:level')
-            logger.log(lvl, 'Executing storm query [%s] as [%s]', text, user)
+            logger.log(lvl, 'Executing storm query {%s} as [%s]', text, user)
```
Comment (vEpiphyte) [inline synapse/cortex.py]: why change this?

File: synapse/cortex.py [hunk 11]
```diff
@@ -102,6 +103,48 @@ def getModelDict(self):
...
+        Return generic information about the cortex including model definition
+        '''
+        return self.cell.getCoreInfo()
+
+    async def addTrigger(self, condition, query, *, info):
+        '''
+        Adds a trigger to the cortex
+        '''
+        username = None if self.user is None else self.user.name
+        self.cell.triggers.add(username, condition, query, info=info)
```
Comment (vEpiphyte) [inline synapse/cortex.py]: Do we want to have a permission (like we have for feed data) about what users may add triggers?

File: synapse/cortex.py [hunk 12]
```diff
@@ -102,6 +103,48 @@ def getModelDict(self):
         '''
         return self.cell.model.getModelDict()
 
+    def getCoreInfo(self):
+        '''
+        Return generic information about the cortex including model definition
+        '''
+        return self.cell.getCoreInfo()
```
Comment (vEpiphyte) [inline synapse/cortex.py]: re-use `stat()` here (see note below)

File: synapse/cortex.py [hunk 13]
```diff
@@ -102,6 +103,48 @@ def getModelDict(self):
...
+        self.cell.triggers.add(username, condition, query, info=info)
+
+    def _trig_auth_check(self, triguser):
+        username = None if self.user is None else self.user.name
+        if username is not None and (not self.user.admin) and username != triguser:
+            raise s_exc.AuthDeny(user=self.user.name, mesg='As non-admin, may only manipulate triggers created by you')
+
+    async def delTrigger(self, iden):
+        '''
+        Deletes aa trigger from the cortex
```
Comment (vEpiphyte) [inline synapse/cortex.py]: spelling

File: synapse/lib/syntax.py [hunk 14]
```diff
@@ -728,13 +749,36 @@ def query(self):
 
             self.ignore(whitespace)
 
-            #if self.nextstr('{'):
-                #self.offs += 1
-                #oper = self.query()
+            # if self.nextstr('{'):
```
Comment (vEpiphyte) [inline synapse/lib/syntax.py]: can we gut this unused subquery logic stub?

File: synapse/lib/syntax.py [hunk 15]
```diff
@@ -1458,7 +1505,30 @@ def quoted(self):
 
             text += c
 
-        self._raiseSytaxError('unexpected end of query text')
+        self._raiseSyntaxError('unexpected end of query text')
+
+    def singlequoted(self):
```
Comment (vEpiphyte) [inline synapse/lib/syntax.py]: this is missing the `\\` escape logic from `quoted()` and is used interchangbly. It also has a nearly the same purpose. Can we just make the quote character for `quoted()` an arg that defaults to `"` which we can override by providing `'` when needed?  That way we don't have two different quoted/escaping rules possible.

File: synapse/lib/trigger.py [hunk 16]
```diff
@@ -1,69 +1,262 @@
...
-        Args:
-            perm ((str,dict)):  The perm tufo to trigger
-            *args (list):       args list to use calling the trigger function
-            **kwargs (dict):    kwargs dict to use calling the trigger function
+        finally:
+            RecursionDepth.reset(token)
 
-        Returns:
-            (None)
+    async def fire(self, node, cond, *, info):
```
Comment (vEpiphyte) [inline synapse/lib/trigger.py]: if we're not a `EventBus` or a `Base` object, do we want to be using the function `fire()` ?  Having seen it used in other code I was expecting the Trigger object to be using the observable pattern.

File: synapse/lib/trigger.py [hunk 17]
```diff
@@ -1,69 +1,262 @@
...
+                    if cond == rule.cond:
+                        await rule.execute(node, tag)
+                if tag[0] != '.' and '.' in tag:
+                    for rule in self._rule_by_tag[glob_tag]:
+                        if cond == rule.cond:
+                            await rule.execute(node, tag)
+
+    def add(self, username, condition, query, *, info):
+
+        iden = os.urandom(16)
```
Comment (vEpiphyte) [inline synapse/lib/trigger.py]: `s_common.guid()`

File: synapse/tests/test_cmds_trigger.py [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+
+class CmdTriggersTest(s_t_utils.SynTest):
+
+    async def test_triggers(self):
+        async with self.getTestDmon('dmoncore') as dmon, await self.agetTestProxy(dmon, 'core') as core:
+            outp = self.getTestOutp()
+            cmdr = await s_cmdr.getItemCmdr(core, outp=outp)
+
+            # Bad storm syntax
+            # self.araises(s_exc.BadStormSyntax, cmdr.runCmdLine('trigger add node:add teststr {[ | | testint=1 ] }'))
```
Comment (vEpiphyte) [inline synapse/tests/test_cmds_trigger.py]: why leave this test commented out?

File: synapse/lib/trigger.py [hunk 19]
```diff
@@ -1,69 +1,262 @@
...
+    class Rule:
+        ver: int  # version: must be 0
+        cond: str  # condition from above list
+        user: str  # username
+        storm: str  # story query
+        form: Optional[str] = dataclasses.field(default=None) # form name
+        tag: Optional[str] = dataclasses.field(default=None) # tag name
+        prop: Optional[str] = dataclasses.field(default=None) # property name
+
+        def __post_init__(self):
```
Comment (vEpiphyte) [inline synapse/lib/trigger.py]: can we use `mesg='...'` syntax for the synerr's we are raising here?

File: synapse/lib/node.py [hunk 20]
```diff
@@ -220,6 +220,10 @@ def getNodeRefs(self):
                 univ = self.snap.model.prop(prop.univ)
                 await univ.wasSet(self, curv)
 
+        await self.snap.core.triggers.fire(self, 'prop:set', info={'prop': f'{self.form.name}.{prop.name}'})
```
Comment (vEpiphyte) [inline synapse/lib/node.py]: Why can't we use prop.full ?

File: synapse/cortex.py (hunk 11 above)
Comment (jnwatson) [reply synapse/cortex.py]: That's valid. I think that's best to kick into a separate task.

Comment (jnwatson) [reply synapse/cortex.py]: (Added)

File: synapse/cortex.py (hunk 12 above)
Comment (jnwatson) [reply synapse/cortex.py]: See above

File: synapse/lib/trigger.py (hunk 16 above)
Comment (jnwatson) [reply synapse/lib/trigger.py]: ok. After trying several alternatives, it is now `run`

File: synapse/tests/test_cmds_trigger.py (hunk 18 above)
Comment (jnwatson) [reply synapse/tests/test_cmds_trigger.py]: Good catch. Fixed it and added it back.

File: synapse/lib/trigger.py (hunk 19 above)
Comment (jnwatson) [reply synapse/lib/trigger.py]: I don't understand.

File: synapse/lib/node.py (hunk 20 above)
Comment (jnwatson) [reply synapse/lib/node.py]: What is that?

File: synapse/cortex.py (hunk 13 above)
Comment (jnwatson) [reply synapse/cortex.py]: ok

File: synapse/lib/node.py [hunk 21]
```diff
@@ -31,7 +31,7 @@ def __init__(self, snap, buid=None, rawprops=None):
         self.props = {}
         self.univs = {}
 
-        #self.vars = {}  # runtime storm variables
+        # self.vars = {}  # runtime storm variables
```
[the remark this answers is not in the document]
Comment (jnwatson) [reply synapse/lib/node.py]: nope. Will remove

File: synapse/lib/syntax.py (hunk 14 above)
Comment (jnwatson) [reply synapse/lib/syntax.py]: yep

File: synapse/lib/trigger.py (hunk 19 above)
Comment (jnwatson) [reply synapse/lib/trigger.py]: got it

File: synapse/lib/node.py (hunk 20 above)
Comment (jnwatson) [reply synapse/lib/node.py]: done

**Pull request merged**